File: native/androidApp/src/main/java/com/reported/nativeandroid/app/ComposerVehicleLookupCoordinator.kt
package com.reported.nativeandroid.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import android.os.SystemClock
import android.util.Log
import com.reported.nativeandroid.BuildConfig
import com.reported.nativeandroid.analytics.ReportedAnalytics
import com.reported.nativeandroid.batch.BatchSubmitStore
import com.reported.nativeandroid.di.AppGraph
import com.reported.nativeandroid.media.LocalSubmissionMediaCleaner
import com.reported.nativeandroid.media.ParseMediaUploader
import com.reported.nativeandroid.remoteconfig.ReportedRemoteConfig
import com.reported.shared.model.AppThemeMode
import com.reported.shared.model.Catalogs
import com.reported.shared.model.CityReportingRules
import com.reported.shared.model.DraftMedia
import com.reported.shared.model.DraftPlateCandidate
import com.reported.shared.model.PhiladelphiaMobilityAccessCatalogs
import com.reported.shared.model.PhiladelphiaMobilityAccessDetails
import com.reported.shared.model.PlatePatternClassifier
import com.reported.shared.model.ReportDraft
import com.reported.shared.model.ReportFilter
import com.reported.shared.model.ReportSummary
import com.reported.shared.model.SubmitReportCommand
import com.reported.shared.model.SubmitReportMediaFile
import com.reported.shared.model.VehicleLookupDetails
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.Job
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.Locale

internal class ComposerVehicleLookupCoordinator(
    private val scope: kotlinx.coroutines.CoroutineScope,
    private val currentState: () -> ComposerUiState,
    private val updateState: ((ComposerUiState) -> ComposerUiState) -> Unit,
    private val onPersistDraft: () -> Unit
) {
    private var vehicleClassificationDebugJob: Job? = null
    private var vehicleDetailsLookupJob: Job? = null

private fun refreshVehicleClassificationDebugNote() {
    if (!BuildConfig.DEBUG) return
    val normalizedPlate = PlatePatternClassifier.normalizePlateInput(currentState().plate)
        .take(PlatePatternClassifier.MAX_LICENSE_PLATE_LENGTH)
    if (
        normalizedPlate.length < 2 ||
        (normalizedPlate.firstOrNull() != 'T' && normalizedPlate.firstOrNull() != 'Y')
    ) {
        applyVehicleClassificationDebugNote(null)
        return
    }

    vehicleClassificationDebugJob?.cancel()
    vehicleClassificationDebugJob = scope.launch {
        delay(350)
        val debugNote = runCatching {
            AppGraph.shared.previewVehicleEnrichmentDebugNoteUseCase.execute(normalizedPlate)
        }.getOrNull()
        val latestPlate = PlatePatternClassifier.normalizePlateInput(currentState().plate)
            .take(PlatePatternClassifier.MAX_LICENSE_PLATE_LENGTH)
        if (latestPlate == normalizedPlate) {
            applyVehicleClassificationDebugNote(debugNote)
        }
    }
}

fun refresh() {
    refreshVehicleDetailsLookup()
    refreshVehicleClassificationDebugNote()
}

private fun refreshVehicleDetailsLookup() {
    val previousDetails = currentState().vehicleLookupDetails
    val lookupPlate = PlatePatternClassifier.normalizePlateInput(currentState().plate)
        .takeIf { it.length in 2..PlatePatternClassifier.MAX_LICENSE_PLATE_LENGTH }
    val lookupState = currentState().plateRegion.trim().uppercase()
        .takeIf { it.matches(Regex("^[A-Z]{2}$")) }
    vehicleDetailsLookupJob?.cancel()
    if (previousDetails != null) {
        updateState { current ->
            current.copy(
                philadelphiaMobilityAccessDetails = current.philadelphiaMobilityAccessDetails
                    .withoutVehicleLookupPrefill(previousDetails)
            )
        }
    }
    if (lookupPlate == null || lookupState == null) {
        updateState {
            it.copy(
                vehicleLookupDetails = null,
                vehicleLookupInFlight = false,
                vehicleLookupMessage = null
            )
        }
        return
    }

    val lookupKey = "$lookupState:$lookupPlate"
    updateState {
        it.copy(
            vehicleLookupDetails = null,
            vehicleLookupInFlight = true,
            vehicleLookupMessage = null
        )
    }
    vehicleDetailsLookupJob = scope.launch {
        delay(500)
        try {
            val details = AppGraph.shared.lookupVehicleDetailsUseCase.execute(lookupPlate, lookupState)
            if (currentState().vehicleLookupKey() != lookupKey) return@launch
            updateState { current ->
                current.copy(
                    vehicleLookupDetails = details,
                    vehicleLookupInFlight = false,
                    vehicleLookupMessage = if (details == null) {
                        "No vehicle details were found for this plate."
                    } else {
                        null
                    },
                    philadelphiaMobilityAccessDetails = details?.let { vehicle ->
                        current.philadelphiaMobilityAccessDetails.copy(
                            vehicleMake = current.philadelphiaMobilityAccessDetails.vehicleMake
                                .ifBlank { vehicle.vehicleMake.orEmpty() },
                            vehicleModel = current.philadelphiaMobilityAccessDetails.vehicleModel
                                .ifBlank { vehicle.vehicleModel.orEmpty() },
                            bodyStyle = current.philadelphiaMobilityAccessDetails.bodyStyle
                                .ifBlank { vehicle.vehicleBody.orEmpty() }
                        )
                    } ?: current.philadelphiaMobilityAccessDetails
                )
            }
            if (details != null) onPersistDraft()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            if (currentState().vehicleLookupKey() != lookupKey) return@launch
            updateState {
                it.copy(
                    vehicleLookupDetails = null,
                    vehicleLookupInFlight = false,
                    vehicleLookupMessage = "Vehicle details are unavailable right now. You can still submit the report."
                )
            }
        }
    }
}

private fun applyVehicleClassificationDebugNote(debugNote: String?) {
    if (!BuildConfig.DEBUG) return
    updateState { current ->
        val nextLines = if (current.notes.isBlank()) {
            mutableListOf()
        } else {
            current.notes
                .lineSequence()
                .filterNot { it.startsWith(VehicleClassificationDebugPrefix) }
                .toMutableList()
        }
        debugNote?.trim()
            ?.takeIf { it.isNotBlank() }
            ?.let(nextLines::add)
        current.copy(notes = nextLines.joinToString("\n"))
    }
    onPersistDraft()
}

}

File: native/iosApp/ReportediOS/Sources/ComposerVehicleLookupCoordinator.swift
import Foundation
import SharedCore

private let vehicleClassificationDebugPrefix = "[DEBUG] Vehicle classification"

@MainActor
final class ComposerVehicleLookupCoordinator {
    typealias StateMutation = (inout ComposerState) -> Void

    private let currentState: () -> ComposerState
    private let updateState: (StateMutation) -> Void
    private let onPersistDraft: () -> Void
    private var classificationDebugTask: Task<Void, Never>?
    private var detailsLookupTask: Task<Void, Never>?

    init(
        currentState: @escaping () -> ComposerState,
        updateState: @escaping (StateMutation) -> Void,
        onPersistDraft: @escaping () -> Void
    ) {
        self.currentState = currentState
        self.updateState = updateState
        self.onPersistDraft = onPersistDraft
    }

    deinit {
        classificationDebugTask?.cancel()
        detailsLookupTask?.cancel()
    }

    func refresh() {
        refreshVehicleDetailsLookup()
        refreshVehicleClassificationDebugNote()
    }

    var currentLookupKey: String? {
        vehicleLookupKey(for: currentState())
    }

    private func refreshVehicleClassificationDebugNote() {
        #if DEBUG
        let normalizedPlate = ComposerViewModel.normalizedPlateInput(currentState().plate)
        guard normalizedPlate.count >= 2,
              normalizedPlate.count <= 10,
              normalizedPlate.hasPrefix("T") || normalizedPlate.hasPrefix("Y") else {
            applyVehicleClassificationDebugNote(nil)
            return
        }
        classificationDebugTask?.cancel()
        classificationDebugTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 350_000_000)
            guard let self, !Task.isCancelled else { return }
            let debugNote = try? await SharedBridge.shared.container.previewVehicleEnrichmentDebugNoteUseCase.execute(
                plate: normalizedPlate
            )
            await MainActor.run { [weak self] in
                guard let self,
                      !Task.isCancelled,
                      ComposerViewModel.normalizedPlateInput(self.currentState().plate) == normalizedPlate else {
                    return
                }
                self.applyVehicleClassificationDebugNote(debugNote)
            }
        }
        #endif
    }

    private func vehicleLookupKey(for state: ComposerState) -> String? {
        let plate = ComposerViewModel.normalizedPlateInput(state.plate)
        let region = state.plateRegion.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard plate.count >= 2,
              plate.count <= 10,
              region.count == 2,
              region.allSatisfy(\.isLetter) else {
            return nil
        }
        return "\(region):\(plate)"
    }

    private func refreshVehicleDetailsLookup() {
        detailsLookupTask?.cancel()
        let snapshot = currentState()
        if let previousDetails = snapshot.vehicleLookupDetails {
            updateState { state in
                state.philadelphiaMobilityAccessDetails = Self.clearingLookupPrefill(
                    state.philadelphiaMobilityAccessDetails,
                    details: previousDetails
                )
            }
        }
        guard let lookupKey = vehicleLookupKey(for: currentState()) else {
            updateState { state in
                state.vehicleLookupDetails = nil
                state.vehicleLookupInFlight = false
                state.vehicleLookupMessage = nil
            }
            return
        }
        let components = lookupKey.split(separator: ":", maxSplits: 1).map(String.init)
        guard components.count == 2 else { return }
        let lookupState = components[0]
        let lookupPlate = components[1]

        updateState { state in
            state.vehicleLookupDetails = nil
            state.vehicleLookupInFlight = true
            state.vehicleLookupMessage = nil
        }
        detailsLookupTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 500_000_000)
            guard let self, !Task.isCancelled else { return }
            do {
                let details = try await SharedBridge.shared.container.lookupVehicleDetailsUseCase.execute(
                    plate: lookupPlate,
                    licenseState: lookupState
                )
                await MainActor.run { [weak self] in
                    guard let self,
                          !Task.isCancelled,
                          self.vehicleLookupKey(for: self.currentState()) == lookupKey else {
                        return
                    }
                    self.updateState { state in
                        state.vehicleLookupDetails = details
                        state.vehicleLookupInFlight = false
                        state.vehicleLookupMessage = nil
                        if let details {
                            state.philadelphiaMobilityAccessDetails = Self.prefilling(
                                state.philadelphiaMobilityAccessDetails,
                                from: details
                            )
                        }
                    }
                    if details != nil {
                        self.onPersistDraft()
                    }
                }
            } catch {
                await MainActor.run { [weak self] in
                    guard let self,
                          !Task.isCancelled,
                          self.vehicleLookupKey(for: self.currentState()) == lookupKey else {
                        return
                    }
                    self.updateState { state in
                        state.vehicleLookupDetails = nil
                        state.vehicleLookupInFlight = false
                        state.vehicleLookupMessage = "Vehicle details are unavailable right now. You can still submit the report."
                    }
                }
            }
        }
    }

    private func applyVehicleClassificationDebugNote(_ debugNote: String?) {
        #if DEBUG
        updateState { state in
            let existingLines = state.notes
                .components(separatedBy: .newlines)
                .filter { !$0.hasPrefix(vehicleClassificationDebugPrefix) }
            let nextLines = existingLines + [debugNote]
                .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            state.notes = nextLines.joined(separator: "\n")
        }
        onPersistDraft()
        #endif
    }

    private static func prefilling(
        _ current: PhiladelphiaMobilityAccessDetails,
        from details: VehicleLookupDetails
    ) -> PhiladelphiaMobilityAccessDetails {
        PhiladelphiaMobilityAccessDetails(
            blockNumber: current.blockNumber,
            streetName: current.streetName,
            zipCode: current.zipCode,
            vehicleMake: current.vehicleMake.isEmpty ? (details.vehicleMake ?? "") : current.vehicleMake,
            vehicleModel: current.vehicleModel.isEmpty ? (details.vehicleModel ?? "") : current.vehicleModel,
            bodyStyle: current.bodyStyle.isEmpty ? (details.vehicleBody ?? "") : current.bodyStyle,
            vehicleColor: current.vehicleColor,
            violationObserved: current.violationObserved,
            frequency: current.frequency
        )
    }

    private static func clearingLookupPrefill(
        _ current: PhiladelphiaMobilityAccessDetails,
        details: VehicleLookupDetails
    ) -> PhiladelphiaMobilityAccessDetails {
        PhiladelphiaMobilityAccessDetails(
            blockNumber: current.blockNumber,
            streetName: current.streetName,
            zipCode: current.zipCode,
            vehicleMake: current.vehicleMake.caseInsensitiveCompare(details.vehicleMake ?? "") == .orderedSame
                ? ""
                : current.vehicleMake,
            vehicleModel: current.vehicleModel.caseInsensitiveCompare(details.vehicleModel ?? "") == .orderedSame
                ? ""
                : current.vehicleModel,
            bodyStyle: current.bodyStyle.caseInsensitiveCompare(details.vehicleBody ?? "") == .orderedSame
                ? ""
                : current.bodyStyle,
            vehicleColor: current.vehicleColor,
            violationObserved: current.violationObserved,
            frequency: current.frequency
        )
    }
}
